scrapy/signalmanager.py

```diff
@@ -3,15 +3,36 @@
 
 from scrapy.utils import signal as _signal
 from scrapy.exceptions import ScrapyDeprecationWarning
-from scrapy.dispatch.robustapply import robustApply
+from scrapy.dispatch.robustapply import robust_apply as _robust_apply
 from scrapy.dispatch.utils import func_accepts_kwargs
-
+from scrapy.dispatch import Signal as _Signal
 
 class SignalManager(object):
 
     def __init__(self, sender=None):
         self.sender = sender
-        self.signals = {}
+        self._patched_receivers = {}
+        self._signal_proxies = {}
+
+    def _ensure_signal(self, signal):
+        """
+        This method is for backward compatability for any custom signals
+        that might be in use in third party extensions which is still using the
+        standard python objects as signals. This method returns a Signal()
+        class instance and saves it to proxy the user defined signal.
+        """
+        if isinstance(signal, _Signal):
+            return signal
+        # Ensure we got at least an old style signal
+        warnings.warn("Signals in scrapy are no longer instances of the basic "
+                      "python object but rather the Signal class defined in "
+                      "the scrapy.dispatch module. Please refer to the Signal "
+                      "API documentation.",
+                      ScrapyDeprecationWarning, stacklevel=3)
+        signal_name = signal.__repr__()
+        if signal_name not in self._signal_proxies:
+            self._signal_proxies.setdefault(signal_name, _Signal())
+        return self._signal_proxies[signal_name]
```

Comment:
1. There is a race condition here, where `signal_name` is not in `self._signal_proxies` at the time when the `if` condition is evaluated, but is already when the next line is evaluated (oh the perks of asynchronous programming). You're already doing the right thing by using `setdefault`, but then you can also skip the "check if already in dict" step.
2. Dictionary keys can be anything in Python, and more importantly, an object's `__repr__` is not necessarily unique (since it could potentially be overwritten and no longer contain the object memory address). You could avoid the `__repr__` detour and just use `signal` itself as key.

Lastly, `setdefault` has the cool quality of returning the value that already was or is now stored in the dictionary, so the four lines above (everything after the warning) could be simplified to a single one:
```python
        return self._signal_proxies.setdefault(signal, _Signal())
```